ostd/src/arch/x86/dispatcher.rs

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{borrow::ToOwned, vec::Vec};
+
+use align_ext::AlignExt;
+
+use crate::{
+    boot::memory_region::MemoryRegionType, device::dispatcher::io_mem::IoMemDispatcherBuilder,
+};
+
+/// Initializes the allocatable MMIO area based on the x86-64 memory distribution map.
+///
+/// In x86-64, the available physical memory area is divided into two regions below 32 bits (Low memory)
+/// and above (High memory). Where the area from the top of low memory to 0xffff_ffff, after the top of
+/// high memory, is the available MMIO area.
+pub(super) fn construct_io_mem_dispatcher_builder() -> IoMemDispatcherBuilder {
+    let mut regions = crate::boot::memory_regions().to_owned();
+    assert!(!regions.is_empty());
+    regions.sort_by_key(|a| a.base());
+    regions.reverse();
+
+    let mut ranges = Vec::new();
+    let mut is_tohm_found = false;
+    const HIGH_MMIO_SIZE: usize = 0x100_0000;
+    const HIGH_MMIO_ALIGN: usize = 0x8_0000_0000;
+
+    for region in regions {
+        if region.typ() != MemoryRegionType::Usable {
+            continue;
+        }
+
+        if !is_tohm_found && (region.base() + region.len()) > u32::MAX as usize {
+            // Find the TOHM (Top of High Memory) and initialize High MMIO region.
+            // Here, using 16MB (0x100_0000) as the size of High MMIO region.
+            //
+            // TODO: Update the High MMIO region in runtime.
+            is_tohm_found = true;
+            // Align start address to HIGH_MMIO_ALIGN
+            let mmio_start_addr = (region.base() + region.len()).align_up(HIGH_MMIO_ALIGN);
+            ranges.push(mmio_start_addr..(mmio_start_addr + HIGH_MMIO_SIZE));
+        } else if (region.base() + region.len()) < u32::MAX as usize {
+            // Find the TOLM (Top of Low Memory) and initialize Low MMIO region (TOLM ~ 0xffff_ffff).
+            // Align start address to 0x1000_0000
+            let mmio_start_addr = (region.base() + region.len()).align_up(0x1000_0000);
+            ranges.push(mmio_start_addr..0x1_0000_0000);
+            break;
+        }
+    }
+
+    if !is_tohm_found {
+        ranges.push(HIGH_MMIO_ALIGN..(HIGH_MMIO_ALIGN + HIGH_MMIO_SIZE));
```

Comment:
The `HIGH_MMIO_ALIGN` comes from the MMIO high address assigned to the VirtIO device, while the `HIGH_MMIO_SIZE` is a configurable magic number.

Reply:
> The `HIGH_MMIO_ALIGN` comes from the MMIO high address assigned to the VirtIO device

But there may be devices other than VirtIO. They may not work?

> while the `HIGH_MMIO_SIZE` is a configurable magic number.

I guess this means that the value is chosen randomly and it "just works", but there are no sources?

Reply:
> But there may be devices other than VirtIO. They may not work?

I think the upper MMIO region is referred to as the "free" MMIO region, in contrast to the fixed MMIO region of the `APIC`. Devices that connect via PCIe can reset the Base Address Register (BAR) to point to this upper MMIO region. 

The VirtIO that I mentioned earlier is used to determine the starting address of the "free" MMIO region.

> I guess this means that the value is chosen randomly and it "just works", but there are no sources?

Unfortunately, I couldn't find sources to support the `HIGH_MMIO_SIZE`. I considered the upper MMIO region as the extendable region and left it as a TODO.

Reply:
Well, the problem is, I don't think that's universally true.

For example, I show the memory map on my laptop below, do you think the algorithm works here?

```
[    0.000000] BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009efff] usable
[    0.000000] BIOS-e820: [mem 0x000000000009f000-0x00000000000fffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x000000003d328fff] usable
[    0.000000] BIOS-e820: [mem 0x000000003d329000-0x000000003dc28fff] reserved
[    0.000000] BIOS-e820: [mem 0x000000003dc29000-0x000000003fa3efff] usable
[    0.000000] BIOS-e820: [mem 0x000000003fa3f000-0x0000000041adefff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000041adf000-0x0000000044a8efff] ACPI NVS
[    0.000000] BIOS-e820: [mem 0x0000000044a8f000-0x0000000044bfefff] ACPI data
[    0.000000] BIOS-e820: [mem 0x0000000044bff000-0x0000000044bfffff] usable
[    0.000000] BIOS-e820: [mem 0x0000000044c00000-0x00000000493fffff] reserved
[    0.000000] BIOS-e820: [mem 0x000000004a000000-0x00000000503fffff] reserved
[    0.000000] BIOS-e820: [mem 0x00000000c0000000-0x00000000cfffffff] reserved
[    0.000000] BIOS-e820: [mem 0x00000000fed20000-0x00000000fed7ffff] reserved
[    0.000000] BIOS-e820: [mem 0x00000000ff000000-0x00000000ffffffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000100000000-0x00000008afbfffff] usable
```

```
[    0.374022] pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
[    0.374023] pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
[    0.374024] pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000bffff window]
[    0.374025] pci_bus 0000:00: root bus resource [mem 0x50400000-0xbfffffff window]
[    0.374025] pci_bus 0000:00: root bus resource [mem 0x4000000000-0x7fffffffff window]
[    0.374026] pci_bus 0000:00: root bus resource [bus 00-fe]
```

I think there are at least two problems:
 - The algorithm here cannot capture the MMIO region below 1MB, i.e. `0x000a0000-0x000bffff`. This contains the location of the VGA ascii-mode framebuffer. So if you want to display something on the screen, you might want to write something here.
 - I don't think the algorithm here can correctly capture the MMIO region above 4GB, i.e. `0x4000000000-0x7fffffff`, either. It seems that the start address is incorrectly assumed to be `0x10_0000_0000` instead. The size is also wrong.

Reply:
Will there be problems if we just treat all holes in the physical memory map as MMIO regions, regardless of where those holes are located? Yes, it will certainly cause some false positives, but:
 - The existing algorithm cannot prevent false positives.
 - I assume doing so won't cause any soundness problems?

Reply:
Thanks for the input! I admit that the current partitioning of IoMem regions is somewhat simplistic, and these regions only support a working version.

The problem appears to be that the peripheral driver cannot access the MMIO regions of the peripheral. We can continuously optimize the range under the security invariant (The peripheral driver cannot access the system device's MMIO) to ensure that all peripheral MMIO areas can be covered.

With the information above, I think a FIXME is deserved in the `construct_io_mem_dispatcher_builder` function to indicate that it may not address all peripheral MMIOs. We should continuously fix (or refine) it in the future.

Reply:
Leaving a FIXME here will quickly break something if you really try to move the PCI bus outside of OSTD. For example, maybe the TDX environment or our Aliyun cloud machine? When's the right time to fix it?

Do you have a response to this question mentioned above?
> Will there be problems if we just treat all holes in the physical memory map as MMIO regions, regardless of where those holes are located?

Reply:
> Leaving a FIXME here will quickly break something if you really try to move the PCI bus outside of OSTD. For example, maybe the TDX environment or our Aliyun cloud machine? When's the right time to fix it?

Looking at the implementation of `VirtAddrAllocator`, I think there are two solutions:
1. We can specify a region large enough to cover all upper MMIOs.
2. We can introduce a field within each `allocator` that specifies whether it is extendable. If peripheral drivers request an MMIO region, such as `0x4_0000_0000` to `0x4_1000_0000`, which exceeds the allocatable area, the dispatcher can identify this situation and dynamically adjust the allocator during runtime.

> Do you have a response to this question mentioned above?
> 
> > Will there be problems if we just treat all holes in the physical memory map as MMIO regions, regardless of where those holes are located?

I looked up memory distribution resources for x86. What I find credible is the information provided by [OSDev](https://wiki.osdev.org/Memory_Map_(x86)). According to the OSDev, the holes we used are "typically" or "possible" used for MMIO.

If we trust the OSDev, then I think there should be no problems if we treat the holes as MMIO regions.

Reply:
As for the VGA framebuffer (`0xa0000` ~ `0xbFFFF`), this place is fixed and can be added to the MMIO dispatcher.

Reply:
It seems that the simplest way is to treat all holes (including a large "hole" at the top) as MMIO regions.

> As for the VGA framebuffer (`0xA0000` ~ `0xBFFFF`), this place is fixed and can be added to the MMIO dispatcher.

Yes, we can do this manually, but using holes in the physical map seems more straightforward, IMO.

I think we can make some immediate fixes in this PR rather than leaving it for future work. Otherwise, it's too wired for me to introduce some known buggy code. However, if you still want to leave it for future PRs, that's less ideal for me, but I can live with it.

Reply:
I adjusted the alignment and the top of the High MMIO region, which now have more reasonable values of `0x1_0000_0000` and `0x8000_0000_0000`, respectively. I think these changes address most cases.

If there are ranges beyond `0x8000_0000_0000`, we should consider dynamically updating the High MMIO region.

Reply:
> I think these changes address most cases.

Yes, these limits seem much more reasonable to me.